sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/container/ThroughputContainerController.java

```diff
@@ -0,0 +1,313 @@
+            })
+            .thenReturn((T) this);
+    }
+
+    private Mono<ThroughputContainerController> resolveDatabaseResourceId() {
+        return this.targetContainer.getDatabase().read()
+            .flatMap(response -> {
+                this.targetDatabaseRid = response.getProperties().getResourceId();
+                return Mono.just(this);
+            });
+    }
+
+    private Mono<ThroughputContainerController> resolveContainerResourceId() {
+        return this.targetContainer.read()
+            .flatMap(response -> {
+                this.targetContainerRid = response.getProperties().getResourceId();
+                return Mono.just(this);
+            });
+    }
+
+    private Mono<ThroughputContainerController> resolveContainerMaxThroughput() {
+        return Mono.defer(() -> Mono.just(this.throughputResolveLevel))
+            .flatMap(throughputResolveLevel -> {
+                if (throughputResolveLevel == ThroughputResolveLevel.CONTAINER) {
+                    return this.resolveThroughputByResourceId(this.targetContainerRid)
+                        .onErrorResume(throwable -> {
+                            if (this.isOfferNotConfiguredException(throwable)) {
+                                this.throughputResolveLevel = ThroughputResolveLevel.DATABASE;
+                            }
+
+                            return Mono.error(throwable);
+                        });
+                } else if (throughputResolveLevel == ThroughputResolveLevel.DATABASE) {
+                    return this.resolveThroughputByResourceId(this.targetDatabaseRid)
+                        .onErrorResume(throwable -> {
+                            if (this.isOfferNotConfiguredException(throwable)) {
+                                this.throughputResolveLevel = ThroughputResolveLevel.CONTAINER;
+                            }
+
+                            return Mono.error(throwable);
+                        });
+                }
+
+                // All the underlying throughput control groups are using target throughput,
+                // which is constant value, hence no need to resolve throughput
+                return Mono.empty();
+            })
+            .flatMap(throughputResponse -> {
+                this.updateMaxContainerThroughput(throughputResponse);
+                return Mono.empty();
+            })
+            .retryWhen(
+                // Throughput can be configured on database level or container level
+                // Retry at most 1 time so we can try on database and container both
+                RetrySpec.max(1).filter(throwable -> this.isOfferNotConfiguredException(throwable))
+            ).thenReturn(this);
+    }
+
+    private Mono<ThroughputResponse> resolveThroughputByResourceId(String resourceId) {
+        checkArgument(StringUtils.isNotEmpty(resourceId), "ResourceId can not be null or empty");
+        return this.client.queryOffers(resourceId, new CosmosQueryRequestOptions())
```

Comment:
This would fail for serverless accounts.. Worth checking wit Thomas Weiss what they want to happen in this case - like is there any way the backend introduced in the meantime how we could retrieve the max throughput allowed for serverless - we don't want to hard-code it in the SDK so we would need some way to get it from the gateway backend to make this work for serverless. Explicitly non-blocking - meaning feel free to merge the PR even without resolving this.

Reply:
Thanks @FabianMeiswinkel , I missed this point completely. 
For now I created an GitHub issue to track it, will sync with Thomas for this one.